src/Database/PostgreSQL/PQTypes/SQL/Builder.hs

```diff
@@ -431,21 +431,33 @@ instance Sqlable SqlSelect where
   toSQLCommand cmd = smconcat
     [ emitClausesSepComma "WITH" $
       map (\(name,command) -> name <+> "AS" <+> parenthesize command) (sqlSelectWith cmd)
+    , if hasUnion then "(" else ""
     , "SELECT" <+> (if sqlSelectDistinct cmd then "DISTINCT" else mempty)
     , sqlConcatComma (sqlSelectResult cmd)
     , emitClause "FROM" (sqlSelectFrom cmd)
     , emitClausesSep "WHERE" "AND" (map toSQLCommand $ sqlSelectWhere cmd)
+    -- If there's a union, the result is sorted and has a limit, applying the
+    -- order and limit to the main subquery won't reduce the overall query
+    -- result, but might reduce its processing time.
+    , if hasUnion && not (null $ sqlSelectOrderBy cmd) && sqlSelectLimit cmd >= 0
+      then smconcat [orderByClause, limitClause]
+      else ""
+    , if hasUnion then ")" else ""
     , emitClausesSep "UNION" "UNION" (sqlSelectUnion cmd)
     , emitClausesSepComma "GROUP BY" (sqlSelectGroupBy cmd)
     , emitClausesSep "HAVING" "AND" (sqlSelectHaving cmd)
-    , emitClausesSepComma "ORDER BY" (sqlSelectOrderBy cmd)
+    , orderByClause
     , if sqlSelectOffset cmd > 0
       then unsafeSQL ("OFFSET " ++ show (sqlSelectOffset cmd))
       else ""
     , if sqlSelectLimit cmd >= 0
-      then unsafeSQL ("LIMIT " ++ show (sqlSelectLimit cmd))
+      then limitClause
       else ""
     ]
+    where
+      hasUnion      = not . null $ sqlSelectUnion cmd
+      orderByClause = emitClausesSepComma "ORDER BY" $ sqlSelectOrderBy cmd
+      limitClause   = unsafeSQL $ "LIMIT" <+> show (sqlSelectLimit cmd)
```

Comment:
We could perhaps make this one a bit easier to understand by
- pulling out the main select clause
- using `parenthesize` (unconditionally) over `if hasUnion then ")" else ""`

Something along these lines (though let-in-where isn't exactly):

```Haskell
instance Sqlable SqlSelect where
  toSQLCommand cmd =
     let mainSelectClause = smconcat
            [ "SELECT" <+> (if sqlSelectDistinct cmd then "DISTINCT" else mempty)
            , sqlConcatComma (sqlSelectResult cmd)
            , emitClause "FROM" (sqlSelectFrom cmd)
            , emitClausesSep "WHERE" "AND" (map toSQLCommand $ sqlSelectWhere cmd)
            -- If there's a union and the result is sorted and has a limit, applying the
            -- order and limit to the main subquery won't reduce the overall query
            -- result, but might reduce its processing time.
            , if hasUnion && not (null $ sqlSelectOrderBy cmd) && sqlSelectLimit cmd >= 0
              then smconcat [orderByClause, limitClause]
              else ""
            ]
    in smconcat
        [ emitClausesSepComma "WITH" $
          map (\(name,command) -> name <+> "AS" <+> parenthesize command) (sqlSelectWith cmd)
        , parenthesize mainSelectClause
        , emitClausesSep "UNION" "UNION" (sqlSelectUnion cmd)
        , emitClausesSepComma "GROUP BY" (sqlSelectGroupBy cmd)
        , emitClausesSep "HAVING" "AND" (sqlSelectHaving cmd)
        , orderByClause
        , if sqlSelectOffset cmd > 0
          then unsafeSQL ("OFFSET " ++ show (sqlSelectOffset cmd))
          else ""
        , if sqlSelectLimit cmd >= 0
          then limitClause
          else   ""
        ]
   where
      hasUnion      = not . null $ sqlSelectUnion cmd
      orderByClause = emitClausesSepComma "ORDER BY" $ sqlSelectOrderBy cmd
      limitClause   = unsafeSQL $ "LIMIT" <+> show (sqlSelectLimit cmd)
```

Reply:
Done.